executor/src/witgen/range_constraints.rs

```diff
@@ -10,77 +11,206 @@ use super::symbolic_evaluator::SymbolicEvaluator;
+    min: T,
+    max: T,
 }
 
 impl<T: FieldElement> RangeConstraint<T> {
+    /// Constraint that allows no higher bits set than the one given
+    /// (counting from zero).
+    pub fn from_max_bit(max_bit: u64) -> Self {
+        Self::from_mask(mask_from_bits::<T>(max_bit + 1))
+    }
+
+    /// Constraint that forces several bits to be set to zero.
+    /// Semantics: x & mask == x
+    pub fn from_mask<M: Into<T::Integer>>(mask: M) -> Self {
+        let mask = mask.into();
+        let max = T::from(cmp::min(mask, (T::from(-1)).to_integer()));
+        Self {
+            mask,
+            min: 0.into(),
+            max,
+        }
+    }
+
+    /// Constraint that only allows this exact value.
+    pub fn from_value(value: T) -> Self {
+        Self {
+            mask: value.to_integer(),
+            min: value,
+            max: value,
+        }
+    }
+
+    /// Constraint that allows values from min to max.
+    /// If min <= max, this means min <= x && x <= max.
+    /// If min > max, this means min <= x || x <= max.
+    pub fn from_range(min: T, max: T) -> Self {
+        let mask = if min <= max {
+            mask_from_bits::<T>(max.to_integer().num_bits() as u64)
+        } else {
+            !T::Integer::from(0)
+        };
+        Self { mask, min, max }
+    }
+
+    /// Returns a bit mask. This might be drastically underfitted in case
+    /// the constraint is more resembling an interval.
+    /// Semantics: X & mask == X holds for all possible values of X.
     pub fn mask(&self) -> &T::Integer {
         &self.mask
     }
-}
 
-impl<T: FieldElement> RangeConstraint<T> {
-    pub fn from_max_bit(max_bit: u64) -> Self {
-        assert!(max_bit < 1024);
-        RangeConstraint {
-            mask: T::Integer::from(((1 << (max_bit + 1)) - 1) as u32),
-        }
+    /// Returns a min-max inclusive range. Note that `max` can be smaller than `min`. In this case the range wraps.
+    /// Semantics: X >= range().0 && X <= range().1 holds for all possible values of X.
+    pub fn range(&self) -> (T, T) {
+        (self.min, self.max)
     }
 
-    pub fn from_mask<M: Into<T::Integer>>(mask: M) -> Self {
-        RangeConstraint { mask: mask.into() }
+    /// Returns the number of field elements included in the range.
+    pub fn range_width(&self) -> T::Integer {
+        range_width(self.min, self.max)
     }
 
     /// The range constraint of the sum of two expressions.
-    pub fn try_combine_sum(&self, other: &RangeConstraint<T>) -> Option<RangeConstraint<T>> {
-        if self.mask & other.mask == 0u32.into() {
-            Some(RangeConstraint {
-                mask: self.mask | other.mask,
-            })
+    pub fn combine_sum(&self, other: &Self) -> Self {
+        let mask = (self.mask & other.mask == 0.into()).then_some(self.mask | other.mask);
+        // TODO can we say anything otherwise?
+        // Something like (self.mask + other.mask) | self.mask | other.mask?
```

Comment:
I think your idea is correct: `mask` is the largest value representable, so the sum will set one extra bit in the most significant side for the carry in every sequence of ones the masks happen to overlap. So, you don't even need to check, and the whole code could be:
```
let mask = (self.mask + other.mask) | self.mask | other.mask;
```

Reply:
*sigh* there is no addition for that type...